mathics/builtin/numbers/numbertheory.py

```diff
@@ -99,8 +99,8 @@ def apply(self, n, evaluation):
         "Divisors[n_Integer]"
         if n == Integer0:
             return None
-        return Expression(
-            SymbolList, *sympy.divisors(n.to_sympy()), element_conversion_fn=from_sympy
+        return to_expression(
```

Comment:
I don't get the `to_expression`

Reply:
If this is just about the fact that this could be Expression with an integer conversion, I'll have to check and you may be right!

However if this is about `to_expression()` in general, begin rant....

What we had been seeing is that way too much time is being spent in the Expression() construction calling from_python() on the elements because we allowed flexible forms like:

```
Expression("Plus", 1, 2)
Expression(SymbolPlus, 1, 2)
Expression(SymbolPlus  1, Integer(2))
```

and so on. While it might not be all that important to use:

```
Expression(SymbolBesselJ, Integer(n), Integer(z))
```

instead of:

```
Expression("BesselJ", n, z)
```

inside the class BesselJ apply (or really eval) method. The fact that Expression has to do the conversion, impacts _other_ calls where it _does_ matter that we don't want to slow down an Expression constructor.

I think if this like traffic jams we sometimes get in NYC. Let's say you are a casual driver who just wants to travel from the suburbs north to a place further out in the suburbs but the route happens to cross one of the few major East-West bridges that daily commuters use to drive into the NYC. Because there is so much traffic on the bridge it now sprawls into the north entrance to the bridge and the road you are traveling on just happens to have an exit that connects to the bridge. So now you are impacted even though you aren't really contributing to the traffic jam.

(I don't know if this analogy helps, but I thought it neat to mention.)


So this is where `to_expression()` comes it. If you want to pass anything other than:

```
Expression(SymbolPlus, Integer1, Integer2)
```

such as the forms shown above, then write this as `to_expression("Plus, 1, 2)` or whatever variation you want.

In contrast to `Expression()` we also allow you to specify a conversion function, so that above could also be written `to_expression("Plus", 1, 2, elements_conversion_fn=Integer)` and that will avoid using the generic `from_python()`. For a list of two items this has very little impact. However for a function an expression like `Range[1000]` we have a bit by telling `to_expression` we are working with integers and needing `from_python()` to figure this out.

(Currently the code doesn't do this, but a small change now will allow this in case you want to do.)

Lastly, because it been annoying and frustrating let me describe a little history.

It was rightly noticed that there were too many `from_python()` calls. So the remedy was to change the order of `if` conditions or suggest that we could inline stuff (if not here then definitely other places) because "function calls are too slow", or that certain forms of function calls are too slow. Or that we should Cythonize everything which from a developer standpoint is a huge time drain waiting for Cython do its thing for some code that will change in a minute or two.

Instead, what I think was and is needed is to understand how the code works at a higher level such as I have tried to describe and understand how we can organize the code so that we don't need conversions all over the place. This and the rewrite/apply pattern matching are making us at least one order of magnitude too slow.

It may be that after the code is reorganized so that things are sorted, `from_python()` will be *slower* because our pattern of use is the opposite: where before most of the time we were asking to convert objects that were already converted, in the future when calls are sorted we will _rarely_ ask to convert something already converted because we call `Expression()` when we don't want conversion and `to_expression()` when we do.

And I am finding something similar in say `N_apply()` where there is this expression customization that has been expanded and probably needs to be reworked. And we have tons of `has_form("List", None)`  or Expression("List", ... ) that we wouldn't need to convert if we had a cleaner base to start out with which understands how to deal with Lists more efficiently

In general we expanded from am messy base and expanding from a clean base makes less work.


Ok. Sorry to rant.